packages/flutter/test/widgets/scroll_position_test.dart

```diff
@@ -0,0 +1,186 @@
+  expect(find.text('10100'), findsNothing, reason: 'with maintainState: $maintainState');
+
+  navigatorKey.currentState.pop();
+  await tester.pump(); // again, navigating always takes two frames
+
+  // Ensure we don't clamp the scroll offset even during the navigation.
+  // https://github.com/flutter/flutter/issues/4883
+  final ScrollableState state = tester.state(find.byType(Scrollable).first);
+  expect(state.position.pixels, equals(1000.0), reason: 'with maintainState: $maintainState');
+
+  await tester.pump(const Duration(seconds: 1));
+
+  // we're 600 pixels high, each item is 100 pixels high, scroll position is
+  // 1000, so we should have exactly 6 items, 10..15.
+
+  expect(find.text('0'), findsNothing, reason: 'with maintainState: $maintainState');
+  expect(find.text('1'), findsNothing, reason: 'with maintainState: $maintainState');
+  expect(find.text('8'), findsNothing, reason: 'with maintainState: $maintainState');
+  expect(find.text('9'), findsNothing, reason: 'with maintainState: $maintainState');
+  expect(find.text('10'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('11'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('12'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('13'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('14'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('15'), findsOneWidget, reason: 'with maintainState: $maintainState');
+  expect(find.text('16'), findsNothing, reason: 'with maintainState: $maintainState');
+  expect(find.text('100'), findsNothing, reason: 'with maintainState: $maintainState');
+}
+
+void main() {
+  testWidgets('whether we remember our scroll position', (WidgetTester tester) async {
+    await performTest(tester, true);
+    await performTest(tester, false);
+  });
+  testWidgets('scroll alignment is honored by ensureVisible', (WidgetTester tester) async {
+    final List<int> items = List<int>.generate(11, (int index) => index).toList();
+    final List<FocusNode> nodes = List<FocusNode>.generate(11, (int index) => FocusNode(debugLabel: 'Item ${index + 1}')).toList();
+    final ScrollController controller = ScrollController();
+    await tester.pumpWidget(
+      MaterialApp(
+        home: ListView(
+          scrollDirection: Axis.vertical,
+          controller: controller,
+          children: items.map<Widget>((int item) {
+            return Focus(
+              key: ValueKey<int>(item),
+              focusNode: nodes[item],
+              child: Container(height: 110),
+            );
+          }).toList(),
+        ),
+      ),
+    );
+
+    controller.position.ensureVisible(tester.renderObject(find.byKey(const ValueKey<int>(0))), alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtStart);
```

Comment:
Nit: Can we break these up to multi-line?

Reply:
Done.